Feat/week-1

tomchaccom opened the pull request:
# 변경점 👍
- 게시글 CRUD 를 JDBC를 통해서 구현
 
# 버그 해결 💊
- Repository의 조회 메소드를 실행했을때 계속 null 값이 반환되었는데 엔티티에 @setter를 적용하니 해결
- JDBC의 RowMapper가 Setter를 기반으로 동작해서 setter가 없으니 값이 들어가지 않아 null로 나왔음

# 테스트 💻
- postman, h2 DB를 통해서 결과를 하나씩 확인합

# 스크린샷 🖼
 ### 게시글 생성하기 

<img width="697" height="433" alt="image" src="https://github.com/user-attachments/assets/bd89735a-d5e1-4dbf-a71d-9836d4fbdbc6" />
<img width="697" height="433" alt="image" src="https://github.com/user-attachments/assets/5ffd86fb-4732-4082-b0ed-6d6f8ca84ea3" />

- 생성한 게시글의 수정 시간은 비어 있고, default로 좋아요 수는 0개로 세팅 

### 게시글 단일 조회
<img width="697" height="515" alt="image" src="https://github.com/user-attachments/assets/ed9e451a-57d3-4eb5-bd57-e4a17d05301b" />

### 게시글 목록 조회
<img width="697" height="675" alt="image" src="https://github.com/user-attachments/assets/1b2ec209-fcb0-4d08-8e82-d5c1dc151c74" />

### 게시글 수정하기
<img width="697" height="388" alt="image" src="https://github.com/user-attachments/assets/43226f98-f4b3-456e-85ee-6f1ec4eb5c79" />

<img width="697" height="388" alt="image" src="https://github.com/user-attachments/assets/a766a646-edcb-43ca-8026-1da37149b348" />

- 수정된 게시글은 수정 시간이 찍혀있음을 확인 가능 

### 게시글 삭제하기 
<img width="697" height="388" alt="image" src="https://github.com/user-attachments/assets/cc390a7f-774e-4fec-8b0f-c8cf3ed2a43d" />

<img width="697" height="388" alt="image" src="https://github.com/user-attachments/assets/3f8094fc-51c5-4aff-88df-9604ea99d63c" />

- 1번 게시글이 삭제되었음을 확인 가능 

 
# 비고 ✏
- JDBC는 JPA가 지원하는 관계 매핑과 같은 @OneToMany, @ManyToOne 은 어떻게 해결하는지 궁금합니다! 
- 실제로 JPA 말고 JDBC로 직접 SQL 쿼리를 작성해야 하는 순간이 있나요?

tomchaccom committed:
Update README.md

tomchaccom committed:
Merge branch 'ApptiveDev:main' into main

Commit:
chore(gradle) : lombok,h2 의존성 추가

Commit:
chore(gradle) : JPA 의존성 추가

Commit:
feat(entity) : 게시글 엔티티 설계

Commit:
chore(gradle) : JPA -> JDBC 의존성 추가 및 삭제

Commit:
fix(entity) : JPA 어노테이션 제거

Commit:
chore(entity): 엔티티 좋아요 수 필드명 수정

Commit:
feat(repo) : JDBC Template를 통해서 CRUD 구현

Commit:
feat(service) : 서비스 계층 개발
- 기존 repo가 findById 시에 int를 인자로 받아 Long으로 수정

Commit:
style(entity) : 가독성 향상을 위한 띄어쓰기

Commit:
feat(controller) : CRUD API 설계 완료

Commit:
feat(h2): h2 설정, 스키마 sql로 작성
- h2 url, 사용자 작성
- JPA와 다르게 JDBC는 직접 DDL을 작성

Commit:
style(sql) : 들여쓰기

Commit:
feat(exception) : 사용자 정의 예외 추가,

Commit:
feat(Post) : 게시판 작성 로직 생성, DTO 및 API 정의

Commit:
feat(Get) : 게시물 단일 조회, 모든 게시글 조회 로직 생성, DTO 및 API 정의

Commit:
feat(Get) : 게시물 수정 로직 생성, DTO 및 API 정의

Commit:
feat(update) : 게시물 수정 로직 생성, DTO 및 API 정의

Commit:
Merge remote-tracking branch 'origin/dev' into dev

Commit:
chore(import) : 의존성 추가

Commit:
fix(save) : SQL 오류 수정 및 수정 시간 오류 해결

sinsehwan commented on src/main/java/com/example/devSns/Post/Post.java:
```diff
@@ -0,0 +1,37 @@
+@AllArgsConstructor
+@NoArgsConstructor
+@Getter
+@Setter
```
Entity에서 Setter 사용은 지양하는 것이 좋습니다! 왜냐하면, setter를 사용한다는 것은 외부에서 entity 값을 마음대로 수정하도록 하는 것인데, 이는 캡슐화 관점에서 좋지 않습니다. 객체의 값을 직접 수정하지 않고 객체에 수정 작업을 시킨다는 관점으로 코드를 작성하면 더 객체 지향적인 코드가 됩니다! entity에서 setter를 사용하지 않도록 습관 들이면 좋습니다!

sinsehwan commented on src/main/java/com/example/devSns/Post/EntityNotFoundException.java:
```diff
@@ -0,0 +1,7 @@
+package com.example.devSns.Post;
+
+public class EntityNotFoundException extends RuntimeException {
+    public EntityNotFoundException(String message) {
+        super(message);
+    }
```
예외 클래스명으로 예외에 대한 맥락을 알 수 있겠네요! 좋습니다

sinsehwan commented on src/main/java/com/example/devSns/Post/Post.java:
```diff
@@ -0,0 +1,37 @@
+    public void Update(UpdatePostRequestDTO DTO){
+        this.content = DTO.content();
+        this.username = DTO.username();
+        this.updatedAt = LocalDateTime.now();
```
저는 자바 환경에서 변수명을 카멜 케이스로 통일하는 편이 협업 시 좋다고 생각합니다! 그래서 dto 또는 requestDto 정도로 변수명을 정하는 것이 통일성 측면에서 좋을 것 같습니다.

sinsehwan commented on src/main/java/com/example/devSns/Post/PostRepository.java:
```diff
@@ -0,0 +1,88 @@
+@Repository
+public class PostRepository {
+
+    private final JdbcTemplate jdbcTemplate;
```
jdbcTemplate를 사용하셨네요! 좋습니다 이후에 JPA로 전환해보면서 ORM이 등장한 배경과 차이를 더 체감해보시면 좋을 것 같습니다!

sinsehwan commented on src/main/resources/schema.sql:
```diff
@@ -0,0 +1,8 @@
+CREATE TABLE IF NOT EXISTS post (
```
잘 작성해 주셨네요! 현재는 개발 중이니 제약조건을 세세하게 걸지 않아도 괜찮지만 추후에는 not null 등 제약조건을 최종적으로 추가해야 할 것 같습니다.

sinsehwan commented on src/main/java/com/example/devSns/Post/PostRepository.java:
```diff
@@ -0,0 +1,88 @@
+                entity.getCreatedAt(),
+                entity.getUpdatedAt()
+        );
+    }
```
추가적으로, Jdbc 기반으로 구현할 때 jdbcTemplate 말고도 jdbcClient를 사용하는 방식도 있습니다! ?(물음표) 대신 파라미터명으로 명시적으로 작성할 수 있다는 장점이 있으니 알아보시면 좋을 것 같습니다!

sinsehwan reviewed (approved):
jdbc 기반으로 과제 또는 프로젝트를 수행해보신 적이 있으신 것 같네요! 코드 구조가 직관적이라서 추후 전환이나 기능 추가 시에도 큰 어려움이 없을 것 같습니다. 다만 변수명 명명규칙이 일관적이지 않아 보여서 통일하는 것이 좋을 것 같습니다! 고생 많으셨습니다

Commit:
chore: 카멜케이스 적용

Commit:
chore : 카멜 케이스 적용

Commit:
fix(JPA) : JDBC에서 JPA로 패러다임 변경
- jpa 의존성 추가
- SQL 로 처리하던 로직을 jpa를 통해 처리

Commit:
feat: 댓글 엔티티 추가

Commit:
feat: 엔티티 연관관계 매핑
- 게시글 : 댓글 일대다
- 대댓글(댓글에 댓글 추가) 일대다

Commit:
feat: 엔티티, 컨트롤러, 서비스, 리포지터리 생성

Commit:
feat(dto) : 댓글 수정 및 생성용 Dto 와 메소드 개발

Commit:
feat(CRUD) : 댓글 CRUD 로직 개발

Commit:
feat(comment) : 댓글 생성 API 개발 완료
- Dto 엔티티 변환 로직을 통해서 댓글에 post 값 할당

Commit:
feat(comment) : 댓글 조회, 수정, 삭제 구현
- 조회시에는 연관된 게시물에 댓글을 조회하도록 설정
- 수정, 삭제 URL은 조금더 고민해보기

Commit:
feat(reply) : 대댓글 작성 로직 및 API 개발

Commit:
fix(comment) : 댓글 조회 API 오류 해결
- 댓글 조회시, 같은 게시글에 달린 대댓글이 중복으로 조회되는 문제 해결
- 댓글 중 parent(부모댓글) 이 null 값인 것들만 조회하도록 수정

Commit:
chore : 사용하지 않는 병수 및 의존성 제거

Commit:
fix(post) : 게시글 조회 API 오류 해결
- 게시글도 댓글과 마찬가지로 조회시, 같은 게시글에 달린 대댓글이 중복으로 조회되는 문제 해결
- 댓글 중 parent(부모댓글) 이 null 값인 것들만 조회하는 메소드를 게시글 service에도 적용

Commit:
chore: 사용되지 않는 메소드 리턴 제거
